File: race-timer-client-web/src/api/RepositoryClient.ts
import { DocumentRequest, DocumentResponse } from '../models/ApiTypes';

/**
 * HTTP client for Race Results Repository REST API.
 * Uses browser fetch API for communication.
 */
export class RepositoryClient {
  private readonly serverUrl: string;
  private readonly jwtToken: string;
  private readonly timeoutMs: number = 10000;

  constructor(serverUrl: string, jwtToken: string) {
    this.serverUrl = serverUrl.replace(/\/$/, ''); // Remove trailing slash
    this.jwtToken = jwtToken;
  }

  /**
   * Get the Start List for a regatta edition (name + start date).
   */
  async getStartList(regattaId: string, regattaStartDate: string): Promise<DocumentResponse | null> {
    try {
      const params = new URLSearchParams({
        regattaId,
        regattaStartDate,
        type: 'START_LIST',
        matchType: 'EXACT',
      });
      const response = await this.fetchWithTimeout(
        `${this.serverUrl}/api/v1/documents/search?${params.toString()}`,
        {
          method: 'GET',
          headers: this.getHeaders(),
        }
      );

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }

      const results = await response.json();
      if (results.results && results.results.length > 0) {
        // Fetch the full document (with model data) by id.
        return await this.getDocument(results.results[0].documentId);
      }
      return null;
    } catch (error) {
      console.error('Failed to get Start List', error);
      throw error;
    }
  }

  /**
   * Get a document by ID.
   */
  async getDocument(documentId: number): Promise<DocumentResponse> {
    try {
      const response = await this.fetchWithTimeout(
        `${this.serverUrl}/api/v1/documents/${documentId}`,
        {
          method: 'GET',
          headers: this.getHeaders(),
        }
      );

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }

      return await response.json();
    } catch (error) {
      console.error(`Failed to get document ${documentId}`, error);
      throw error;
    }
  }

  /**
   * Create a new document.
   */
  async createDocument(request: DocumentRequest): Promise<DocumentResponse> {
    try {
      const response = await this.fetchWithTimeout(
        `${this.serverUrl}/api/v1/documents`,
        {
          method: 'POST',
          headers: this.getHeaders(),
          body: JSON.stringify(request),
        }
      );

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Failed to create document', error);
      throw error;
    }
  }

  /**
   * Update an existing document (creates a new version).
   *
   * The server expects the raw model bytes as the request body
   * (application/octet-stream), with changeDescription and format supplied as
   * query parameters. modelData is a base64-encoded string here and is decoded
   * to bytes before being sent.
   */
  async updateDocument(
    documentId: number,
    modelData: string,
    changeDescription: string,
    format: string = 'XMI'
  ): Promise<DocumentResponse> {
    try {
      const params = new URLSearchParams({
        changeDescription,
        format,
      });
      const response = await this.fetchWithTimeout(
        `${this.serverUrl}/api/v1/documents/${documentId}?${params.toString()}`,
        {
          method: 'PUT',
          headers: this.getBinaryHeaders(),
          body: this.base64ToBytes(modelData),
        }
      );

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }

      return await response.json();
    } catch (error) {
      console.error(`Failed to update document ${documentId}`, error);
      throw error;
    }
  }

  /**
   * Check if the server is reachable.
   */
  async isServerReachable(): Promise<boolean> {
    try {
      const response = await this.fetchWithTimeout(
        `${this.serverUrl}/actuator/health`,
        {
          method: 'GET',
          headers: this.getHeaders(),
        }
      );

      return response.ok;
    } catch (error) {
      return false;
    }
  }

  /**
   * Get standard headers for JSON API requests.
   */
  private getHeaders(): HeadersInit {
    return {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${this.jwtToken}`,
    };
  }

  /**
   * Get headers for requests that send raw binary model data.
   */
  private getBinaryHeaders(): HeadersInit {
    return {
      'Content-Type': 'application/octet-stream',
      'Authorization': `Bearer ${this.jwtToken}`,
    };
  }

  /**
   * Decode a base64-encoded string into raw bytes for an octet-stream body.
   */
  private base64ToBytes(base64: string): Uint8Array {
    const binary = atob(base64);
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) {
      bytes[i] = binary.charCodeAt(i);
    }
    return bytes;
  }

  /**
   * Fetch with timeout support.
   */
  private async fetchWithTimeout(
    url: string,
    options: RequestInit
  ): Promise<Response> {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeoutMs);

    try {
      const response = await fetch(url, {
        ...options,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);
      return response;
    } catch (error) {
      clearTimeout(timeoutId);

      if (error instanceof Error && error.name === 'AbortError') {
        throw new Error(`Request timeout after ${this.timeoutMs}ms`);
      }

      throw error;
    }
  }
}

File: race-timer-client/src/main/java/org/rowtown/rms/rrr/client/api/RepositoryClient.java
package org.rowtown.rms.rrr.client.api;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import lombok.extern.slf4j.Slf4j;
import okhttp3.*;

import java.io.IOException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * HTTP client for communicating with the Race Results Repository.
 */
@Slf4j
public class RepositoryClient {

    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");

    private final String baseUrl;
    private final OkHttpClient httpClient;
    private final ObjectMapper objectMapper;
    private String jwtToken;

    public RepositoryClient(String baseUrl, String jwtToken) {
        this.baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
        this.jwtToken = jwtToken;
        this.objectMapper = new ObjectMapper().registerModule(new JavaTimeModule());

        this.httpClient = new OkHttpClient.Builder()
            .connectTimeout(Duration.ofSeconds(30))
            .readTimeout(Duration.ofSeconds(60))
            .writeTimeout(Duration.ofSeconds(60))
            .build();
    }

    /**
     * Set JWT token for authentication.
     */
    public void setJwtToken(String jwtToken) {
        this.jwtToken = jwtToken;
    }

    /**
     * Get the Start List for a regatta edition (name + start date).
     *
     * @param regattaId the regatta name
     * @param regattaStartDate the regatta start date (ISO-8601, e.g. 2024-05-17)
     */
    public DocumentResponse getStartList(String regattaId, String regattaStartDate) throws IOException {
        String url = baseUrl + "/api/v1/documents/search"
                     + "?regattaId=" + enc(regattaId)
                     + "&regattaStartDate=" + enc(regattaStartDate)
                     + "&type=START_LIST&matchType=EXACT";

        Request request = new Request.Builder()
            .url(url)
            .addHeader("Authorization", "Bearer " + jwtToken)
            .get()
            .build();

        try (Response response = httpClient.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                throw new IOException("Failed to get start list: " + response);
            }

            SearchResults results = objectMapper.readValue(
                response.body().string(), SearchResults.class);

            if (results.results != null && !results.results.isEmpty()) {
                // Get full document with model data
                Long docId = results.results.get(0).documentId;
                return getDocument(docId);
            }

            return null;
        }
    }

    /**
     * Get the Regatta Definition (RML) for a regatta edition (name + start date).
     *
     * @param regattaId the regatta name
     * @param regattaStartDate the regatta start date (ISO-8601, e.g. 2024-05-17)
     */
    public DocumentResponse getRegattaDefinition(String regattaId, String regattaStartDate) throws IOException {
        String url = baseUrl + "/api/v1/documents/search"
                     + "?regattaId=" + enc(regattaId)
                     + "&regattaStartDate=" + enc(regattaStartDate)
                     + "&type=RML&matchType=EXACT";

        Request request = new Request.Builder()
            .url(url)
            .addHeader("Authorization", "Bearer " + jwtToken)
            .get()
            .build();

        try (Response response = httpClient.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                throw new IOException("Failed to get regatta definition: " + response);
            }

            SearchResults results = objectMapper.readValue(
                response.body().string(), SearchResults.class);

            if (results.results != null && !results.results.isEmpty()) {
                return getDocument(results.results.get(0).documentId);
            }

            return null;
        }
    }

    /**
     * Delete a document by ID.
     */
    public void deleteDocument(Long documentId) throws IOException {
        String url = baseUrl + "/api/v1/documents/" + documentId;

        Request request = new Request.Builder()
            .url(url)
            .addHeader("Authorization", "Bearer " + jwtToken)
            .delete()
            .build();

        try (Response response = httpClient.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                throw new IOException("Failed to delete document: " + response.code() + " " + response.message());
            }
        }
    }

    /**
     * Get a document by ID.
     */
    public DocumentResponse getDocument(Long documentId) throws IOException {
        String url = baseUrl + "/api/v1/documents/" + documentId;

        Request request = new Request.Builder()
            .url(url)
            .addHeader("Authorization", "Bearer " + jwtToken)
            .get()
            .build();

        try (Response response = httpClient.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                throw new IOException("Failed to get document: " + response);
            }

            return objectMapper.readValue(response.body().string(), DocumentResponse.class);
        }
    }

    /**
     * Create a new document.
     */
    public DocumentResponse createDocument(DocumentRequest documentRequest) throws IOException {
        String url = baseUrl + "/api/v1/documents";
        String json = objectMapper.writeValueAsString(documentRequest);

        RequestBody body = RequestBody.create(json, JSON);
        Request request = new Request.Builder()
            .url(url)
            .addHeader("Authorization", "Bearer " + jwtToken)
            .post(body)
            .build();

        try (Response response = httpClient.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                throw new IOException("Failed to create document: " + response.code() + " " + response.message());
            }

            return objectMapper.readValue(response.body().string(), DocumentResponse.class);
        }
    }

    /**
     * Update a document (creates new version).
     */
    public DocumentResponse updateDocument(Long documentId, byte[] modelData, String changeDescription) throws IOException {
        String url = baseUrl + "/api/v1/documents/" + documentId +
                     "?changeDescription=" + changeDescription +
                     "&format=JSON";

        RequestBody body = RequestBody.create(modelData, MediaType.get("application/octet-stream"));
        Request request = new Request.Builder()
            .url(url)
            .addHeader("Authorization", "Bearer " + jwtToken)
            .put(body)
            .build();

        try (Response response = httpClient.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                throw new IOException("Failed to update document: " + response.code() + " " + response.message());
            }

            return objectMapper.readValue(response.body().string(), DocumentResponse.class);
        }
    }

    /**
     * Check if server is reachable.
     */
    public boolean isServerReachable() {
        try {
            String url = baseUrl + "/api/v1/documents/search?page=0&size=1";
            Request request = new Request.Builder()
                .url(url)
                .addHeader("Authorization", "Bearer " + jwtToken)
                .get()
                .build();

            try (Response response = httpClient.newCall(request).execute()) {
                return response.isSuccessful();
            }
        } catch (IOException e) {
            log.debug("Server not reachable: {}", e.getMessage());
            return false;
        }
    }

    private static String enc(String value) {
        return URLEncoder.encode(value == null ? "" : value, StandardCharsets.UTF_8);
    }

    /**
     * Simple DTOs for API communication.
     *
     * <p>{@code timer} carries the redundancy role: one of "PRIMARY",
     * "FIRST_BACKUP", "SECOND_BACKUP". {@code regattaStartDate} is ISO-8601
     * (yyyy-MM-dd). {@code raceId} is derived server-side and returned in
     * responses only (not sent in requests).</p>
     */
    public static class DocumentResponse {
        public Long documentId;
        public String type;
        public String regattaId;
        public String regattaStartDate;
        public String raceId;
        public String milestoneId;
        public String timer;
        public String author;
        public String description;
        public Long latestVersion;
        public byte[] modelData;
    }

    public static class DocumentRequest {
        public String type;
        public String regattaId;
        public String regattaStartDate;
        public String milestoneId;
        public String timer;
        public String author;
        public String description;
        public Set<String> tags;
        public Map<String, String> metadata;
        public byte[] modelData;
    }

    public static class SearchResults {
        public Long total;
        public Integer page;
        public Integer pageSize;
        public List<DocumentSummary> results;
    }

    public static class DocumentSummary {
        public Long documentId;
        public String type;
        public String regattaId;
        public String regattaStartDate;
        public String raceId;
        public String timer;
        public Long latestVersion;
    }
}
